clients/drcachesim/tracer/raw2trace.cpp

```diff
@@ -878,6 +861,103 @@ raw2trace_t::do_conversion()
+            return nullptr;
+        }
+        hashtable_add(&decode_cache, decode_pc, desc);
+        ret = desc;
+    } else {
+        // TODO(mtrofin): log some rendering of the instruction summary that will be
+        // returned
+        *pc = ret->next_pc();
+    }
+    return ret;
+}
+
+bool
+instr_summary_t::construct(void *dcontext, INOUT app_pc *pc, app_pc orig_pc,
+                           OUT instr_summary_t *desc, uint verbosity)
+{
+    struct instr_destroy_t {
+        instr_destroy_t(void *dcontext_in, instr_t *instr_in)
+            : dcontext(dcontext_in)
+            , instr(instr_in)
+        {
+        }
+        void *dcontext;
+        instr_t *instr;
+        ~instr_destroy_t()
+        {
+            instr_destroy(dcontext, instr);
+        }
+    };
+
+    instr_t *instr = instr_create(dcontext);
+    instr_destroy_t instr_collector(dcontext, instr);
+
+    *pc = decode(dcontext, *pc, instr);
+    if (*pc == nullptr || !instr_valid(instr)) {
+        return false;
+    }
+    DO_VERBOSE(3, {
+        instr_set_translation(instr, orig_pc);
+        dr_print_instr(dcontext, STDOUT, instr, "");
+    });
+    desc->next_pc_ = *pc;
+    desc->packed_ = 0;
+
+    bool is_prefetch = instr_is_prefetch(instr);
+    bool reads_memory = instr_reads_memory(instr);
+    bool writes_memory = instr_writes_memory(instr);
+
+    desc->packed_ |= reads_memory << kReadsMemPos;
+    desc->packed_ |= writes_memory << kWritesMemPos;
+    desc->packed_ |= is_prefetch << kIsPrefetchPos;
+    desc->packed_ |= instru_t::instr_is_flush(instr) << kIsFlushPos;
+    desc->packed_ |= instr_is_cti(instr) << kIsCtiPos;
+
+    desc->type_ = instru_t::instr_to_instr_type(instr);
+    desc->prefetch_type_ = is_prefetch ? instru_t::instr_to_prefetch_type(instr) : 0;
+    desc->length_ = static_cast<byte>(instr_length(dcontext, instr));
+
+    if (reads_memory || writes_memory) {
+#define TRAVERSE_SRCS(WHAT_TO_DO)                            \
+    for (int i = 0, e = instr_num_srcs(instr); i < e; ++i) { \
+        opnd_t op = instr_get_src(instr, i);                 \
+        if (opnd_is_memory_reference(op)) {                  \
+            WHAT_TO_DO;                                      \
+        }                                                    \
+    }
+
+#define TRAVERSE_DSTS(WHAT_TO_DO)                            \
+    for (int i = 0, e = instr_num_dsts(instr); i < e; ++i) { \
+        opnd_t op = instr_get_dst(instr, i);                 \
+        if (opnd_is_memory_reference(op)) {                  \
+            WHAT_TO_DO;                                      \
+        }                                                    \
+    }
+
+        TRAVERSE_SRCS(++desc->srcs_)
+        TRAVERSE_DSTS(++desc->dests_)
+        desc->srcs_and_dests_.resize(desc->srcs_ + desc->dests_);
+        int pos = 0;
+        TRAVERSE_SRCS(desc->srcs_and_dests_[pos++] = op)
+        TRAVERSE_DSTS(desc->srcs_and_dests_[pos++] = op)
```

Comment:
Simpler code that just does a loop with push_back and no macros will be more readable and maintainable, and I wouldn't think any slight overhead here would matter: unless there's profiling evidence I'd rather not have macros and less readable code on a non-critical-path like this.

Reply:
true, and we can always rework this if it pops on any perf radar. reworked.